Update Actions and Repo, and Add New Actions

erogluorhan opened the pull request:
NOTE: Huge number of files changed is due to `pre-commit hook` that is newly setup and ran on all the files in the repo.

This PR is aimed at making this repo look as much applicable/possible as GeoCAT-comp repo with respect to actions, folder structure, etc. for consistency purposes. Therefore it makes the following changes:

- Please first beware that the Github actions "build_test" is now renamed to "CI/Python (...)". Thus, for review/approval, see if the newly named actions are passing even though the older name is required by branch protection rules.


- Adds automated `ci.yml` and removes manually-triggered `build_test.yml`.

- Renames CI action and step names similar to geocat-comp wherever applicable.

- Adds `pre-commit hooks` and removes 'yapf.yaml':
Note: To use these pre-commit hooks in local, each developer needs to do the followings (after pulling this branch or `main` branch after merge into local):
   `pip install pre-commit`
   `pre-commit install`

- Adds a LICENSE file and then several badges into README.md to have similar badges with GeoCAT-comp.

- Adds code coverage reports, `codecov`, as in GeoCAT-comp.

- Adds an action for uploading GeoCAT-f2py tp PyPi.

erogluorhan commented:
Hey @andersy005 , I am adding code coverage to this repo, too. Reports are being generated correctly, I also activated codecov webhooks; however, there is still an issue with"enabling the GeoCAt-f2py repo" in my codecov account. Here is the message:

![image](https://user-images.githubusercontent.com/32553057/112695899-76fdca00-8e4a-11eb-9f37-6a08ab758fdf.png)

The "installation settings" link there is this URL: https://github.com/organizations/NCAR/settings/installations/591816

I think I don't have an Admin status in NCAR organization, and it may be the reason I cannot see above URL and access to NCAR' s organization settings. 

Do you have an idea how to enable GeoCAT-f2py as you did previously for GeoCAT-comp?

andersy005 commented:
> Do you have an idea how to enable GeoCAT-f2py as you did previously for GeoCAT-comp? 

Since this repository is public, I don't think you have to do anything else. I can confirm that the reports are publicly accessible: https://codecov.io/gh/NCAR/geocat-f2py/tree/a987975b7b69532dbc050607a1b1c08ecbe8ee8a 

So, I'd say that everything is working as expected.

andersy005 commented on .github/codecov.yml:
```diff
@@ -0,0 +1,21 @@
+codecov:
+  max_report_age: off
+
+comment: false
+
+coverage:
+  precision: 2
+  round: down
+  status:
+    project:
+      default:
+        threshold: 0.2
+        if_not_found: success
+    patch:
+      default:
+        enabled: no
+        if_not_found: success
+    changes:
+      default:
+        enabled: no
+        if_not_found: success
```
This file shouldn't be here. It should be in the root directory.

erogluorhan replied:
> This file shouldn't be here. It should be in the root directory.

I don't think that's an issue since (1) [codecov docs](https://docs.codecov.io/docs/codecov-yaml) do not explicitly say it should be in the root directory, (2) GeoCAT-comp also has that file under `.github` folder and its badge works fine:

![image](https://user-images.githubusercontent.com/32553057/112698013-60597200-8e4e-11eb-9635-0b6042ac0a1b.png)


But, when you check README of this branch here (`update_actions`), it cannot show the report outputs:

![image](https://user-images.githubusercontent.com/32553057/112698105-939c0100-8e4e-11eb-8123-d395a4a9e8a3.png)

I am checking the codecov badge and URL, and they seem like good, too.

erogluorhan replied:
Ok, I now realize there is nothing to be concerned here since the same warning (i.e. "repo is not enabled") is also being given for GeoCAT-comp but its overview is correctly populated from `main` branch. Once this PR is merged, the badge here as well as the overview at codecov should be correctly populated. 

Thanks for discussing!

pilotchute commented:
Was preparing to give my review, when I saw that some automated function had modified the fortran files. While I do not expect that any of the fortran files have been broken, I do not like that they have been touched at all. They need to remain untouched, in order to preserve our confidence that they do in fact provide the exact desired behavior from NCL.

erogluorhan commented:
> Was preparing to give my review, when I saw that some automated function had modified the fortran files. While I do not expect that any of the fortran files have been broken, I do not like that they have been touched at all. They need to remain untouched, in order to preserve our confidence that they do in fact provide the exact desired behavior from NCL.

This is a great catch, thanks! I totally agree with this.

It was the `pre-commit hooks` that were run on all the files and I missed the files under `src/geocat/f2py/fortran` were also changed. 

I fixed it now by reverting all of those files as well as adding an `exclude` key to all of our pre-commit hooks for this repo(see [.pre-commit-config.yaml)](https://github.com/NCAR/geocat-f2py/blob/update_actions/.pre-commit-config.yaml) as follows:

`exclude: src/geocat/f2py/fortran/`

I need your input here: With this exclusion, we are ignoring not only Fortran files but also all the files such as `.pyf` we implemented under `src/geocat/f2py/fortran/` directory. My thought is that we can/should only exclude Fortran files. What do you think?

pilotchute commented:
`.pyf` files are not a python file type, and the linter shouldn't touch them. I am happy to exclude all of the files in `.../fortran/` since those are only used for compilation during the build phase, and they should remain untouched once tested and approved.

erogluorhan commented:
> `.pyf` files are not a python file type, and the linter shouldn't touch them. I am happy to exclude all of the files in `.../fortran/` since those are only used for compilation during the build phase, and they should remain untouched once tested and approved.

Agreed!

pilotchute commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,37 @@
+    rev: v0.30.0
+    hooks:
+    - id: yapf
+      args: ['--style=.style.yapf', '--recursive', '--in-place']
```
This is a different style than we've been using, I would suggest remaining on the google style.

pilotchute replied:
Nevermind, this is referncing a style file which is pulling google style. all good

pilotchute commented on fast_test.py:
```diff
@@ -1,12 +1,14 @@
 # THIS IS A DEVELOPMENT PERFORMANCE TESTING CLASS, REMOVE WHEN NO LONGER NEEDED
```
This file should not be included in the commit, it isn't an official part of the repo, I don't know how it got into main.

erogluorhan replied:
Removed.

pilotchute commented on setup.py:
```diff
@@ -15,16 +21,42 @@ def version():
+      long_description=long_description,
+      long_description_content_type='text/markdown',
+      classifiers=[
+          'Operating System :: OS Independent',
```
Not true, we are not OS independent for f2py, at least yet. We don't have a build process for windows.

erogluorhan replied:
Great catch (of stealing from GeoCAT-comp :) ). Corrected to MacOS X and Linux now.

pilotchute commented on src/geocat/f2py/__init__.py:
```diff
@@ -1,10 +1,12 @@
-from .rcm2rgrid_wrapper import (rcm2rgrid, rgrid2rcm)
-from .triple_to_grid_wrapper import (grid_to_triple, triple_to_grid,
-                                     grid2triple, triple2grid)
+from .linint2_wrapper import linint1, linint2, linint2_points, linint2pts
```
What is removing the grouping parenthesis? From a style perspective they are quite useful as they allow longer import lists to be multiline.

erogluorhan replied:
It is again pre-commit hook `isort` (i.e. "import sort"). Actually there is a nice behavior of it on these kind of imports  which I really like and would like to share with you:

When it finds multiple imports in a single line where number of chars exceed 79 (or whatever), it puts automatic parentheses and split that line into two or more lines. However, if you just put parentheses into where multiple imports can already fit into a single line, it then removes parentheses. In other words, parentheses are only used for cases where imports do not fit into a single line, which I like. 

Thoughts?

pilotchute replied:
That's fine, I don't like the un-needed modifications (which I think hurt readability), but I'll not worry about it as long as multiline imports won't be broken.

pilotchute commented on src/geocat/f2py/checks.py:
```diff
@@ -1,6 +1,6 @@
-import xarray as xr
-import numpy as np
 import dask as da
+import numpy as np
+import xarray as xr
```
What is reordering the imports?

erogluorhan replied:
It is the pre-commit hook `isort` (i.e. "import sort"). See inside [.pre-commit-config.yaml](https://github.com/NCAR/geocat-f2py/blob/update_actions/.pre-commit-config.yaml) 

```
-   repo: https://github.com/PyCQA/isort    # To sort the imports
    rev: 5.7.0
    hooks:
    - id: isort
      exclude: src/geocat/f2py/fortran/
```

pilotchute replied:
thanks resolveing

pilotchute commented on src/geocat/f2py/rcm2points_wrapper.py:
```diff
@@ -75,19 +79,34 @@ def rcm2points(lat2d, lon2d, fi, lat1d, lon1d, opt=0, msg=None, meta=False):
     Returns
     -------
 
+<<<<<<< HEAD
+        :class:`numpy.ndarray`: The interpolated grid. A multi-dimensional array
+        of the same size as fi except that the rightmost dimension sizes have been
+        replaced by the number of coordinate pairs (lat1dPoints, lon1dPoints).
+        Double if fi is double, otherwise float.
+=======
     	:class:`numpy.ndarray`: The interpolated grid. A multi-dimensional array
     	of the same size as fi except that the rightmost dimension sizes have been
     	replaced by the number of coordinate pairs (lat1dPoints, lon1dPoints).
     	Double if fi is double, otherwise float.
+>>>>>>> 09ab9753fcf8a6f478ce210b770c3643bea0d84e
 
     Description
     -----------
 
+<<<<<<< HEAD
+        Interpolates data on a curvilinear grid, such as those used by the RCM (Regional Climate Model),
+        WRF (Weather Research and Forecasting) and NARR (North American Regional Reanalysis)
+        models/datasets to an unstructured grid. All of these have latitudes that are oriented south-to-north.
+        A inverse distance squared algorithm is used to perform the interpolation.
+        Missing values are allowed and no extrapolation is performed.
+=======
```
This looks like a merge conflict got stuck into the docstring.

erogluorhan replied:
Fixed, thanks!

pilotchute commented on src/geocat/f2py/rcm2rgrid_wrapper.py:
```diff
@@ -164,10 +163,10 @@ def rcm2rgrid(lat2d, lon2d, fi, lat1d, lon1d, msg=None, meta=False):
-            #     fi.dims[-1]: lon2d,
-            #     fi.dims[-2]: lat2d,
-            # },
+        # coords={
+        #     fi.dims[-1]: lon2d,
+        #     fi.dims[-2]: lat2d,
+        # },
```
I disagree with this action by the linter.

erogluorhan replied:
I agree! It was due to a strange behavior of `yapf`, and I fixed it now. FYI: Below is why there was such a behavior:

When we were handling the Google style for `yapf` from inside the style file (`based_on_style = google` in `.style.yapf` in the root directory), there was that weird behavior you pointed out. When I removed that setting from style file and instead put it into `.pre-commit-config.yaml` as an argument of `yapf`  (i.e. `args: ['--style=.style.yapf', '--in-place', '--recursive', '--style', 'google']`), it fixed the thing.

I am still keeping the `yapf` style file, because even though there are a few style configs in it for now, there are many [YAPF Knobs](https://github.com/google/yapf#knobs) that we would be useful in the future.

pilotchute replied:
Thanks that's great.

pilotchute commented on src/geocat/f2py/triple_to_grid_wrapper.py:
```diff
@@ -361,10 +361,10 @@ def triple_to_grid(data,
-            #     data.dims[-1]: x_in,
-            #     data.dims[-2]: y_in,
-            # },
+        # coords={
+        #     data.dims[-1]: x_in,
+        #     data.dims[-2]: y_in,
+        # },
```
another linter modification I don't agree with.
Why is Yapf behaving differently with this implementation? it should still be using the google style that we already had right?

erogluorhan replied:
See above resolution.

pilotchute replied:
resolved

pilotchute commented on test/test_moc_globe_atl.py:
```diff
@@ -65,70 +72,64 @@ class Test_Moc_Globe_Atl(ut.TestCase):
-            xr.DataArray(tmp_a.astype(np.float64)),
-            xr.DataArray(tmp_a.astype(np.float64)), xr.DataArray(t_lat),
-            xr.DataArray(rmlak))
+        out_arr = moc_globe_atl(xr.DataArray(lat_aux_grid),
+                                xr.DataArray(tmp_a.astype(np.float64)),
+                                xr.DataArray(tmp_a.astype(np.float64)),
+                                xr.DataArray(tmp_a.astype(np.float64)),
+                                xr.DataArray(t_lat), xr.DataArray(rmlak))
```
Another linter change I don't like. Prev version was cleaner and more compact.

pilotchute replied:
OBE

erogluorhan replied:
Here is how it looked like in `main` branch:

```
        out_arr = geocat.f2py.moc_globe_atl(
            xr.DataArray(lat_aux_grid), xr.DataArray(tmp_a.astype(np.float64)),
            xr.DataArray(tmp_a.astype(np.float64)),
            xr.DataArray(tmp_a.astype(np.float64)), xr.DataArray(t_lat),
            xr.DataArray(rmlak))
```

I don't agree that it was cleaner and compact because it did not look consistent as the rest of the code (e.g. Lines 90-94) had the function arguments just aligned to each other:

```
out_arr = geocat.f2py.moc_globe_atl(lat_aux_grid,
                                    tmp_a.astype(np.float32),
                                    tmp_a.astype(np.float32),
                                    tmp_a.astype(np.float32), t_lat,
                                    rmlak)
```

This is more of a personal preferences difference though, and I could be fine with either way.

pilotchute commented on test/test_rcm2points.py:
```diff
@@ -68,131 +74,97 @@
-                                   lat,
-                                   lon,
-                                   opt=2))
+            rcm2points(lat2d, lon2d, fi_nom.astype(np.float64), lat, lon,
+                       opt=2))
```
I do like this linter change. We should talk about all the differences we're seeing between the two styles that should have been the same that we are using.

pilotchute replied:
OBE

erogluorhan replied:
> I do like this linter change. We should talk about all the differences we're seeing between the two styles that should have been the same that we are using.

Now, I spent more time on it, and here are my findings:

There is no difference between the behaviors of our older YAPF config or that in this PR. The main difference is that we called `geocat.f2py.rcm2points` before, but now we converted it to `rcm2points` plus an initial `import geocat.f2py`. That being said, I ran both pre-commit yapf and our older yapf (i.e. `python3 -m yapf --in-place --recursive --style google .` of the `pip3`-installed yapf), and both made the exact same changes as follows:

```
rcm2points(lat2d, lon2d, fi_nom.astype(np.float64), lat, lon,
           opt=2))
```

```
geocat.f2py.rcm2points(lat2d, 
                       lon2d, 
                       fi_nom.astype(np.float64), 
                       lat, 
                       lon,
                       opt=2))
```

It looks strange to me :)

pilotchute reviewed (commented):
Files reviewed. I think we still have a conversation ahead about some of the automated changes.

erogluorhan commented:
> Files reviewed. I think we still have a conversation ahead about some of the automated changes.

Thanks for your through review!

I have addressed all of your comments, so I just requested your re-review. Though, I am happy to have a quick chat to talk about any unclear points.

pilotchute commented:
> > Files reviewed. I think we still have a conversation ahead about some of the automated changes.
> 
> Thanks for your thorough review!
> 
> I have addressed all of your comments, so I just requested your re-review. Though, I am happy to have a quick chat to talk about any unclear points.

Thanks, I'll have my review done soon.

erogluorhan commented:
@pilotchute  FYI: I removed style file now since it was almost empty with its current shape and there was no need to keep it in repo. If/when we need to employ more yapf knobs, we can add again. If you think otherwise, let me know.
